### PR TITLE
Add ports and hostname correctly in kube yaml

### PR DESCRIPTION
**Pull request by umohnani8**
If a pod is created with ipc sharing, allow adding
separate ports for each container to the kube yaml
and also set the pod level hostname correctly.

Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

Fixes https://github.com/containers/podman/issues/13030

#### Does this PR introduce a user-facing change?

```release-note
Set ports and hostname correctly in the generated kube yaml.

```

### CODE REVIEW

File: libpod/kube.go [hunk 1]
```diff
@@ -376,17 +385,19 @@ func (p *Pod) podWithContainers(ctx context.Context, containers []*Container, po
...
-			// We add the original port declarations from the libpod infra container
-			// to the first kubernetes container description because otherwise we loose
-			// the original container/port bindings.
-			// Add the port configuration to the first regular container or the first
-			// init container if only init containers have been created in the pod.
-			if first && len(ports) > 0 && (!isInit || len(containers) == 2) {
-				ctr.Ports = ports
-				first = false
+			// infra container, wipe them here only if we are not sharing the ipc namespace
+			if !p.SharesIPC() {
```
Comment (Luap99) [inline libpod/kube.go:389]: Why is this ipc namespace? This looks like it should check for the netns. Ports can only be used when there is a netns, this is not related to the ipc namespace.

Comment (umohnani8) [reply libpod/kube.go:389]: I think we need to come to a consensus on when we allow containers to expose different ports in a pod.
We currently error out when `--share` is set to uts, net, and the default value(ipc,net,uts).
```
➜  podman git:(kube-hostname) ✗ ./bin/podman pod create --name mypod                                       
=====s.NetNS.NSMode: slirp4netns
4fda91cfcc0e6cf5cd48e2ec2f81649a9217d5e6cb6b77e3ec2cc5ffe4802176
➜  podman git:(kube-hostname) ✗ ./bin/podman run --name ss_02  --pod mypod --hostname test -p 5000:6000 -d fedora

=====s.NetNS.NSMode: pod
Error: invalid config provided: published or exposed ports must be defined when the pod is created: network cannot be configured when it is shared with a pod
➜  podman git:(kube-hostname) ✗ ./bin/podman pod create --name mypod --share uts                                 
=====s.NetNS.NSMode: slirp4netns
2293b04ed355ed2338cc1883dd3917bdaec2abbb05cb6f1258a9d9ff150d9603
➜  podman git:(kube-hostname) ✗ ./bin/podman run --name ss_02  --pod mypod --hostname test -p 5000:6000 -d fedora

=====s.NetNS.NSMode: slirp4netns
Error: invalid config provided: cannot set hostname when joining the pod UTS namespace: invalid configuration
➜  podman git:(kube-hostname) ✗ ./bin/podman pod create --name mypod --share net                                 
=====s.NetNS.NSMode: slirp4netns
f2b5169bc3d3599e0110009bee534d73b3e732108ce3161452c73ae7474ac249
➜  podman git:(kube-hostname) ✗ ./bin/podman run --name ss_02  --pod mypod --hostname test -p 5000:6000 -d fedora

=====s.NetNS.NSMode: pod
Error: invalid config provided: published or exposed ports must be defined when the pod is created: network cannot be configured when it is shared with a pod
```

But it works fine with pid and ipc
```
➜  podman git:(kube-hostname) ✗ ./bin/podman pod create --name mypod --share pid                                 
=====s.NetNS.NSMode: slirp4netns
f4c4a3ab9f9ad58e6037fd9db6cee170a01e4dc9d3bb3ec272799968742485ec
➜  podman git:(kube-hostname) ✗ ./bin/podman run --name ss_02  --pod mypod --hostname test -p 5000:6000 -d fedora

=====s.NetNS.NSMode: slirp4netns
c97df4f323dc13ec368798d381ba694685f6b999c5f77113260c18c9ef716eb2
➜  podman git:(kube-hostname) ✗ ./bin/podman pod create --name mypod --share ipc                                 
=====s.NetNS.NSMode: slirp4netns
ee955b9ef17d8039e414cb9375897fe9c5baab3d0389b231191e3eca5fe14bdd
➜  podman git:(kube-hostname) ✗ ./bin/podman run --name ss_02  --pod mypod --hostname test -p 5000:6000 -d fedora

=====s.NetNS.NSMode: slirp4netns
ce3aa72735d21a06aae21eb58bc06725e4d3f379e0e4e73dfd81ea41058432d3
```

Should we always require ports to be defined at the pod level only regardless of which kernel namespaces the pod will share?

Comment (rhatdan) [reply libpod/kube.go:389]: We should match kubernetes,  Can I have two different containers in a Pod listen on differnet ports blocked to the other container.

Comment (Luap99) [reply libpod/kube.go:389]: that all depends on whenever the netns is shared or not, if there is no shared netns then you could add ports to the pod(infra container) but since no container uses the namespace they will not have the same ports. Thefore ports must be per container.

If the netns is shared the ports should only be set on the pod(infra container).

Either way this has nothing to do with the ipc namespace.

Comment (umohnani8) [reply libpod/kube.go:389]: Containers in a kubernetes pod share the network and ipc namespaces by default. You can set separate ports for each container in a pod, it is actually recommended to do that to separate the incoming traffic to each container.

It seems like podman pods has a requirement of defining all the ports at pod creation to associate them with the infra container and a patch was introduced to ensure that. But from testing, I noticed that we only error out when certain kernel namespaces are shared and not on others. 

@Luap99 so basically, we should only enforce port being added at pod creation when only `netns` is shared. If any other namespace is shared, it should be fine to allow separate ports for each container?

cc @mheon

Comment (Luap99) [reply libpod/kube.go:389]: yes hostname is uts, ports is net

Comment (umohnani8) [reply libpod/kube.go:389]: But in kubernetes, hostname is set at the pod level. There is no option to set it for containers. So for that case, should we always add a hostname at the pod level when generating the kube yaml if it is specified for one of the containers in the pod? 
We can also block that option off and only allow hostname to be set when the pod is created regardless of which namespace is being shared.

Comment (mheon) [reply libpod/kube.go:389]: Blocking `generate kube` (or at least warning of unpredictable results) on containers without full namespace sharing might be the most sane thing we can do

Comment (umohnani8) [reply libpod/kube.go:389]: If a pod doesn't have a hostname defined, by default the pod's name is set to the hostname, but it isn't shown in the kube yaml. So we don't need to do anything if a hostname is not defined. 

If we allow users to set hostname for containers when the uts namespace is not being shared, won't it be confusing for them when they go generate a kube yaml for it and don't see a hostname there? 

I know podman pods are slightly different than kubernetes pods, but it might make generate kube easier if we try to follow the similar rules for the two.

Comment (umohnani8) [reply libpod/kube.go:389]: So by default, containers in a kubernetes pod share the same ipc and network namespaces. They can also share the same pid namespace when `v1.PodSpec.shareProcessNamespace: true` is set in the pod spec.

So if a different combination of namespaces are shared we log a warning that the generated kube yaml might not be complete?

Comment (umohnani8) [reply libpod/kube.go:389]: One more thing to note is, podman pod's default is different. We share uts, net, and ipc by default.

Comment (rhatdan) [reply libpod/kube.go:389]: Yes.

Comment (rhatdan) [reply libpod/kube.go:389]: Just make sure this warning does not happen by default, just if users muck around with --share flag.

Comment (rhatdan) [reply libpod/kube.go:389]: Doesn't kubernetes share uts, net and ipc by default? They also share User Namespace by defaut, but don't mention that...

Comment (umohnani8) [reply libpod/kube.go:389]: Yup, that is true my bad. I didn't see it in their docs. Alright I will log a warning if anyone messes with `--share` and will add the restriction to `--hostname`. Thanks!

**Comment by rhatdan** [comment]
@umohnani8 Any movment on this or do you want to hand this off to an intern?

**Comment by umohnani8** [comment]
Made the requested changes and tests are basically green.
@containers/podman-maintainers PTAL

**Comment by TomSweeneyRedHat** [comment]
LGTM
but would like a head nod from @Luap99
